### PR TITLE
Upgrade libsodium to 1.0.18

### LINKED ISSUE #2872: Figure out how to safely upgrade to libsodium 1.0.16+
libsodium 1.0.16 introduces changes to Ed25519 and Curve25519 point validation: https://github.com/jedisct1/libsodium/issues/662 . These could be used to provoke a chain fork if we were to naively upgrade. Some of the changes are disabled if ``ED25519_COMPAT`` is set; some are not.

### PR DESCRIPTION
**Pull request by str4d**
Includes patches that maintain consensus compatibility with libsodium 1.0.15 for Ed25519 pubkey and signature validation.

Replaces #4239. Closes #2872.

### CODE REVIEW

**Comment by str4d** [comment]
Note to reviewers: the patches to pubkey validation and signature validation are applied separately, because the pubkey validation patch on its own fixes all except one of the test case failures, while the signature validation patch is necessary to resolve the last failure, and proves that the set of valid signatures was altered in libsodium 1.0.16.

**Comment by ebfull** [comment]
ACK. I also ran the tests on `master` without upgrading libsodium. Only concern might be the vagueness of the name `test_consensus.cpp` as plenty of other tests are of consensus-critical components as well.

File: depends/patches/libsodium/1.0.15-signature-validation.diff [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
++        /* p (order 4) */
++        { 0xb4, 0x17, 0x6a, 0x70, 0x3d, 0x4d, 0xd8, 0x4f, 0xba, 0x3c, 0x0b,
++          0x76, 0x0d, 0x10, 0x67, 0x0f, 0x2a, 0x20, 0x53, 0xfa, 0x2c, 0x39,
++          0xcc, 0xc6, 0x4e, 0xc7, 0xfd, 0x77, 0x92, 0xac, 0x03, 0xfa },
++        /* p+1 (order 1) */
++        { 0xec, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
++          0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
++          0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x7f },
++        /* p+2707385501144840649318225287225658788936804267575313519463743609750303402022
++           (order 8) */
```
Comment (daira) [inline depends/patches/libsodium/1.0.15-signature-validation.diff]: Please delete all the comments from this array. Some of them are wrong (for example this one)!

**Comment by str4d** [comment]
Addressed @daira's comments.

**Comment by daira** [comment]
Once I've updated the spec I'd like someone to check consistency with this PR, but that doesn't need to block merging it.

**Comment by str4d** [comment]
Force-pushed to fix the commit comment. No change to PR.

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
Blocked on the new dependency being uploaded to the downloads server.

**Comment by mdr0id** [comment]
@zkbot retry

**Pull request merged**